arangod/Cluster/ClusterInfo.cpp

```diff
@@ -1573,6 +1610,125 @@ Result ClusterInfo::createDatabaseCoordinator(  // create database
+
+  // Instruct the Agency to enter the creation of the new database
+  // by entering it into Plan/Databases/ but with the fields
+  // isBuilding: true, and coordinator and rebootId set to our respective
+  // id and rebootId
+  VPackBuilder builder;
+  buildIsBuildingSlice(database, builder);
+
+  AgencyWriteTransaction trx(
+      {AgencyOperation("Plan/Databases/" + database.getName(),
+                       AgencyValueOperationType::SET, builder.slice()),
+       AgencyOperation("Plan/Version", AgencySimpleOperationType::INCREMENT_OP)},
+      AgencyPrecondition("Plan/Databases/" + database.getName(),
+                         AgencyPrecondition::Type::EMPTY, true));
+
+  // TODO: Should this never timeout?
+  res = ac.sendTransactionWithFailover(trx, 0.0);
+
+  if (!res.successful()) {
+    if (res._statusCode == (int)arangodb::rest::ResponseCode::PRECONDITION_FAILED) {
+      return Result(TRI_ERROR_ARANGO_DUPLICATE_NAME);
+    }
+
+    return Result(TRI_ERROR_CLUSTER_COULD_NOT_CREATE_DATABASE_IN_PLAN);
+  }
+
+  // Now update our own cache of planned databases:
+  loadPlan();
+
+  // And wait for our database to show up in `Current/Databases`
+  auto waitresult = waitForDatabaseInCurrent(database);
+
+  if (waitresult.fail()) {
+    // cleanup: remove database from plan?
+    auto ret = cancelCreateDatabaseCoordinator(database);
+
+    if (ret.ok()) {
+      // Creation failed
+      return Result(TRI_ERROR_CLUSTER_COULD_NOT_CREATE_DATABASE,
+                    "database creation failed");
+    } else {
+      // Cleanup failed too
+      return ret;
+    }
+  }
+  return Result();
+}
+
+// Finalize creation of database in cluster by removing isBuilding, coordinator, and coordinatorRebootId;
+// as precondition that the entry we put in createIsBuildingDatabaseCoordinator is still in Plan/ unchanged.
+Result ClusterInfo::createFinalizeDatabaseCoordinator(methods::CreateDatabaseInfo const& database) {
+  AgencyComm ac;
+
+  VPackBuilder pcBuilder;
+  buildIsBuildingSlice(database, pcBuilder);
+
+  VPackBuilder entryBuilder;
+  buildFinalSlice(database, entryBuilder);
+
+  AgencyWriteTransaction trx(
+      {AgencyOperation("Plan/Databases/" + database.getName(),
+                       AgencyValueOperationType::SET, entryBuilder.slice()),
+       AgencyOperation("Plan/Version", AgencySimpleOperationType::INCREMENT_OP)},
+      AgencyPrecondition("Plan/Databases/" + database.getName(),
+                         AgencyPrecondition::Type::VALUE, pcBuilder.slice()));
+
+  auto res = ac.sendTransactionWithFailover(trx, 0.0);
+
+  if (!res.successful()) {
+    if (res._statusCode == (int)arangodb::rest::ResponseCode::PRECONDITION_FAILED) {
+      return Result(TRI_ERROR_CLUSTER_COULD_NOT_CREATE_DATABASE,
+                    "Could not finish creation of database: Plan/Databases/ "
+                    "entry was modified in Agency");
+    }
+    // Something else went wrong.
+    return Result(TRI_ERROR_CLUSTER_COULD_NOT_CREATE_DATABASE);
```

Comment:
This is a similar case as above. The database will not be cleaned up, either, and the coordinator keeps going. Maybe we need a retry loop here, too?

Reply:
Or we at least try to cancel the operation...